src/pip/_internal/commands/install.py

```diff
@@ -594,6 +594,69 @@ def get_lib_location_guesses(*args, **kwargs):
     return [scheme['purelib'], scheme['platlib']]
 
 
+def site_packages_writable(**kwargs):
+    return all(
+        test_writable_dir(d) for d in set(get_lib_location_guesses(**kwargs))
+    )
+
+
+def decide_user_install(
+    use_user_site,  # type: Optional[bool]
+    prefix_path=None,  # type: Optional[str]
+    target_dir=None,  # type: Optional[str]
+    root_path=None,  # type: Optional[str]
+    isolated_mode=False,  # type: bool
+):
+    # type: (...) -> bool
+    """Determine whether to do a user install based on the input options.
+
+    If use_user_site is False, no additional checks are done.
+    If use_user_site is True, it is checked for compatibility with other
+    options.
+    If use_user_site is None, the default behaviour depends on the environment,
+    which is provided by the other arguments.
+    """
+    if use_user_site is False:
+        logger.debug("Non-user install by explicit request")
+        return False
+
+    if use_user_site is True:
+        if prefix_path:
+            raise CommandError(
+                "Can not combine '--user' and '--prefix' as they imply "
+                "different installation locations"
+            )
+        if virtualenv_no_global():
+            raise InstallationError(
+                "Can not perform a '--user' install. User site-packages "
+                "are not visible in this virtualenv."
+            )
+        logger.debug("User install by explicit request")
+        return True
+
+    # If we are here, user installs have not been explicitly requested/avoided
+    assert use_user_site is None
+
+    # user install incompatible with --prefix/--target
+    if prefix_path or target_dir:
+        logger.debug("Non-user install due to --prefix or --target option")
+        return False
+
+    # If user installs are not enabled, choose a non-user install
+    if not site.ENABLE_USER_SITE:
+        logger.debug("Non-user install because user site-packages disabled")
+        return False
+
+    # If we don't have permission for a non-user install, choose a user install
```

Comment:
nit: this comment position feels strange regarding the following `if` that checks the opposite...

Maybe:
```
# Choose non-user install if it is writable, or user install otherwise
if site_packages_writable(root=root_path, isolated=isolated_mode):
```

Reply:
Yep, the code changed without the comment being updated. I've pushed a rewording.